server/hotrod/src/main/scala/org/infinispan/server/hotrod/KeyValueVersionConverterFactory.scala

```diff
@@ -23,19 +22,39 @@ object KeyValueVersionConverterFactory {
          val capacity = UnsignedNumeric.sizeUnsignedInt(key.length) + key.length +
             (if (newValue != null) UnsignedNumeric.sizeUnsignedInt(newValue.length) + newValue.length + 8 else 0)
 
-         val out = ByteBuffer.allocate(capacity)
-         UnsignedNumeric.writeUnsignedInt(out, key.length)
-         out.put(key)
+         val out = Array.ofDim[Byte](capacity)
+         var offset = UnsignedNumeric.writeUnsignedInt(out, 0, key.length)
+         offset = putBytes(key, offset, out)
          if (newValue != null) {
-            UnsignedNumeric.writeUnsignedInt(out, newValue.length)
-            out.put(newValue)
-            out.putLong(newMetadata.version().asInstanceOf[NumericVersion].getVersion)
+            offset = UnsignedNumeric.writeUnsignedInt(out, offset, newValue.length)
+            offset = putBytes(newValue, offset, out)
+            putLong(newMetadata.version().asInstanceOf[NumericVersion].getVersion, offset, out)
          }
-
-         out.array()
+         out
       }
    }
 
+   private def putBytes(bytes: Array[Byte], offset: Int, out: Array[Byte]): Int = {
+      var localOffset = offset
+      bytes.foreach(b => {
+         out.update(localOffset, b)
+         localOffset += 1
+      })
+      localOffset
+   }
+
+   private def putLong(l: Long, offset: Int, out: Array[Byte]): Int = {
```

Comment:
Can the version be negative?  Wondering why we didn't use UnsignedNumeric.writeUnsignedLong.  Granted we can't change it now since it is part of the protocol spec.